### DISCUSSION TITLE
RAC & RA side navigation suggestions

### DISCUSSION THREAD

**Discussion by BMCwebdev** [adobe/react-spectrum#6539] [Q&A]
Hi everyone. I am starting a task to build a side navigation, much like the one on the React Spectrum website. I am wondering if anyone can suggest any ideas or share experience building one using React Aria. I am looking at what is in the utilities and other options. Thought maybe I could use Section, but that seems to only want to live inside of specific components. FocusScope might be useful. There is no `ul` component, and I don’t think I can or want to use the Collections pattern on a `ul` based Link navigation component. I don’t think I want to use any of the overlay components, because they unmount, and the navigation should probably stay in the DOM. 

Well, if anyone has any thoughts I’d love and appreciate hearing them. Thanks!

**Comment by BMCwebdev** [comment]
I'm pretty much going with the idea that I should just roll with my own `<nav>` `<ul>` and `<li>`. ListBox of course was interesting to look at. The whole useListBox documentation page. I could use that to render the semantic markup I am looking for. But not sure if it is all overkill anyways. I don't need selection, I don't need arrow key support (not even sure if I would want that at all).

Yea, think I'm going to just go the traditional route, but again, if anyone has any ideas lemme know!

**Reply by mryechkin** [reply]
I'd be interested to see what you come up with. Currently looking at implementing something very similar as well, and I was hoping to utilize existing RAC components and utilities for most of it.

I was really excited when I saw [submenu](https://react-spectrum.adobe.com/react-aria/Menu.html#submenus) support added to the [Menu](https://react-spectrum.adobe.com/react-aria/Menu.html), but unfortunately it doesn't look like it works all that well when forcing it to stay in the open state. Seems like it's mostly aimed at contextual-style menus, where there's a trigger button, rather than an "always open" navigational menu.

**Reply by mryechkin** [reply]
Here's what I tried.

When I force it to stay open via the `isOpen` prop, the menu seems to lose focus, unless there's a focusable `Button` present as a child of `MenuTrigger`: 

```jsx
import { Menu, MenuTrigger, Popover, SubmenuTrigger } from 'react-aria-components';

// ...

<MenuTrigger isOpen>
  <Menu items={items} {...props}>
    {function renderSubmenu(item) {
      if (item.children) {
        return (
          <SubmenuTrigger>
            <NavItem key={item.name}>{item.name}</NavItem>
            <Popover>
              <PopoverWrapper>
                <Menu items={item.children}>
                  {(subItem) => renderSubmenu(subItem)}
                </Menu>
              </PopoverWrapper>
            </Popover>
          </SubmenuTrigger>
        );
      }
      return <NavItem key={item.name}>{item.name}</NavItem>;
    }}
  </Menu>
</MenuTrigger>
```

Thing is, I don't need a button to trigger the menu, because the menu should just always be open. But I _do_ need to be able to navigate the menu with a keyboard, and for that I need to be able to tab into it. Which seems like it's not possible without a `Button` as a child.

So now I'm not sure if this is even the right approach.

I should mention that for my use case keyboard support with directional arrows is a requirement, hence why I was really hoping to achieve this with existing React Aria components or hooks.

Otherwise, I agree that a simple `<nav>` with a `<ul>` would suffice, and I'd probably go the same route if I didn't need directional arrow support.

**Reply by snowystinger** [reply]
An always open MenuTrigger is going to cause a number of focus issues as it is opening a Dialog/Popover which is a focus trap.

You could probably build this without too much difficulty using [FocusScope](https://react-spectrum.adobe.com/react-aria/FocusScope.html#focusscope) and a [FocusManager](https://react-spectrum.adobe.com/react-aria/FocusScope.html#focusmanager-interface) with a `nav` and `ul` setup.

Otherwise, we'll have a TreeView component soon which could also handle something like this potentially.

**Reply by mryechkin** [reply]
Thanks for the quick reply @snowystinger.

Yeah - I had figured as much, hence my hesitation to continue going down that path. I had thought about using `GridList` for this as well, but the `TreeView` sounds like it would be more appropriate. Any ideas on when that might be available?

My main issue here with rolling my own is that I'd like to have directional keys support, and with just `FocusScope` and plain `nav` and `ul` I'd have to wire all that up myself.

**Reply by snowystinger** [reply]
https://github.com/adobe/react-spectrum/issues/2396

**Reply by BMCwebdev** [reply]
I don't think I am making a navigation like you are if you are considering the Menu component. I wanted a navigation that is always in the DOM, and pretty much looks very similar to what the React Aria website uses. I don't need arrow support because it is a traditional `<nav>` that uses tab. The things I found and tried to use were FocusScope, useKeyboard, useInteractOutside, usePreventScroll. But I have had some issues with them.

FocusScope seems to not let you shift-tab back to the triggering button that is outside of the container. Even with contain set to false. The design I am working with would require that. I still might use it, but only if the benefit outweighs having to do something tricky, like having two buttons in the trigger area.

useKeyboard is working so far for detecting escape key and was a nice quick win.

useInteractOutside I was having race conditions with other states for setting the sidenav open and closed on mobile. So I had to drop it.

usePreventScroll has been a nice quick win and I am still using that.

**Reply by BMCwebdev** [reply]
A follow up question: Would it be possible to allow FocusScope to shift-tab focus back out of its container to the triggering element? I’m not sure why it seems the trigger is removed from being able to tab back to, though once you’ve tabbed back to the root and tab forward again you can focus on it.

**Reply by snowystinger** [reply]
I don't follow. Maybe an example would help?
I can tab in and out of the FocusScope:
https://codesandbox.io/p/sandbox/boring-wave-4663w2?file=%2Fsrc%2FApp.js%3A12%2C23

**Reply by BMCwebdev** [reply]
Hey @snowystinger , you bet, and thanks. I'll just use the RAC demo for FocusScope, but I'll add in a couple of buttons so that we can watch where focus moves to.

I assume there is a reason to do this to avoid a focus trap of some sort, but I was wondering if it still makes sense to take the triggering button out of the tab flow if you are not using `contain` on FocusScope.

In the example, if you click the open button, and then try to shift-tab back to the open button, it will skip over it and go to the "next in reverse shift-tab focus order" item. If you start forward tabbing again, the open button is focusable.

Use case would be if this open button was a toggle, and became the close button after clicking it to open.

https://codesandbox.io/p/sandbox/elated-cache-cnl8lz?file=%2Fsrc%2FApp.js%3A12%2C29

**Reply by snowystinger** [reply]
ah, i see, you're running into https://github.com/adobe/react-spectrum/pull/2459

**Reply by BMCwebdev** [reply]
Ah, ye ol' 2459! 😄 
Thanks @snowystinger
